src/Autumn/compile.jl

```diff
@@ -2,24 +2,306 @@
+        deepcopy(state)
+      end
+     end
+     [initFunction, nextFunction]
+  end
+
+  # initialize state::STATE variable
+  function compileinitstate(data)
+    initStateParamsInternal = map(expr -> :(Dict{Int64, $(haskey(data["types"], expr) ? compile(data["types"][expr]) : Any)}()), 
+    filter(x -> !(x[1] in data["externalVars"]), data["historyVars"]))
+    initStateParamsExternal = map(expr -> :(Dict{Int64, Union{$(compile(data["types"][expr])), Nothing}}()), 
+        filter(x -> (x[1] in data["externalVars"]), data["historyVars"]))
+    initStateParams = [0, initStateParamsInternal..., initStateParamsExternal...]
+    initState = :(global state = STATE($(initStateParams...)))
+    initState
+  end
+  # ----- END HELPER FUNCTIONS ----- #
+
+
+  # ----- COMPILATION -----#
+  if (aexpr.head == :program)
+    # handle AExpression lines
+    lines = filter(x -> x !== :(), map(arg -> compile(arg, aexpr), aexpr.args))
+    
+    # construct STATE struct
+    stateParamsInternal = map(expr -> :($(Symbol(string(expr[1]) * "History"))::Dict{Int64, $(haskey(data["types"], expr) ? compile(data["types"][expr]) : Any)}), 
+                              filter(x -> !(x[1] in data["externalVars"]), data["historyVars"]))
+    stateParamsExternal = map(expr -> :($(Symbol(string(expr[1]) * "History"))::Dict{Int64, Union{$(compile(data["types"][expr])), Nothing}}), 
+                              filter(x -> (x[1] in data["externalVars"]), data["historyVars"]))
+    stateStruct = quote
+      mutable struct STATE
+        time::Int
+        $(stateParamsInternal...)
+        $(stateParamsExternal...)
+      end
+    end
+    initStateStruct = compileinitstate(data)
+    
+    # handle initnext
+    initnextFunctions = compileinitnext(data)
+    prevFunctions = compileprevfuncs(data)
+    builtinFunctions = compilebuiltin()
+
+    # remove empty lines
+    lines = filter(x -> x != :(), 
+            vcat(builtinFunctions, lines, stateStruct, initStateStruct, prevFunctions, initnextFunctions))
+
+    # construct module
+    expr = quote
+      module CompiledProgram
+        export init, next
+        import Base.min
+        using Distributions
+        using MLStyle: @match 
+        $(lines...)
+      end
+    end  
+    expr.head = :toplevel
+    striplines(expr)
+  else
+    throw(AutumnCompileError())
+  end
 end
 
 "Run `prog` forever"
-function runprogram(prog::AProgram)
-  initexterns = (x = 3, y = 2)
-  state = init(p, initexterns)
-  while true
-    externs = (x = rand(3:10), y = rand(1:10))
-    state = next(state, externs)
-  end
-  state
+function runprogram(prog::Expr, n::Int)
+  mod = eval(prog)
+  particles = mod.init(mod.Click(5, 5))
+
+  externals = [nothing, nothing]
+  for i in 1:n
+    externals = [nothing, mod.Click(rand([1:10;]), rand([1:10;]))]
+    particles = mod.next(mod.next(externals[rand(Categorical([0.7, 0.3]))]))
+  end
+  particles
+end
```

Comment:
Perhaps this should be a test?

Reply:
This was a function you had in the original compile.jl template, I'm thinking of modifying it to encapsulate the memory leak/processes stuff in the next PR.